src/docs/emitter-redesign-proposal.md

```diff
@@ -0,0 +1,333 @@
+    - Sets a flag to skip pInvoke result postprocessing - we need to assign a raw payload to the instance.
+
+2. ReturnableMethodHandler (Static, Instance) - Creates a step which takes the return value of the pInvoke and returns it to the caller
+    - StaticMethodHandler - Adds static keyword to the method signature
+    - InstanceMethodHandler - Adds `SwiftSelf` to the pInvoke signature, adds SwiftSelf creation to the marshalling phase (this should call the `ProvideSwiftSelfCreationCode()` method on the type representation) and pushes an appropriate argument to the pInvoke signature and call.
+
+3. SwiftErrorHandler - Adds `SwiftError` to the pInvoke signature and call. Pushes appropriate check to the result postprocessing step.
+
+4. GenericParameterHandler - Adds generic parameters to the method signature and pInvoke signature. Into marshalling phase it adds obtaining the metadata pointers and PWT pointers. Updates the pInvoke signature and the pInvoke call.
+
+5. AsyncMethodHandler - TODO
+
+#### Group 2 - ReturnType Handlers
+
+1. IndirectResultHandler - Creates an indirect result in marshalling phase. Changes the pInvoke signature to return a void. Pushes `IndirectResult` argument to the pInvoke call. Updates the result postprocessing step to marshal the result from Swift.
+
+2. BoundGenericResultHandler - Changes the pInvoke signature to return a buffer struct. Adds postprocessing step to create the bound generic struct from the buffer.
+
+3. DirectResultHandler - Sets the pInvoke signature return type to the original direct result.
+
+4. VoidResultHandler - Sets the pInvoke signature return type to void. Updates the pInvoke call to not return anything. (This can be merged with the DirectResultHandler and delegated to the emitter).
+
+#### Group 3 - Argument Handlers
+
+1. NonFrozenArgumentHandler - Updates the marshalling phase to create a handle for the argument. Pushes the handle to the pInvoke signature and call.
+
+2. GenericArgumentHandler - Updates the marshalling phase to create a handle for the argument. Pushes the handle to the pInvoke signature and call.
+
+3. BoundGenericArgumentHandler - Updates the marshalling phase to cast the argument to a buffer strut. Pushes the buffer struct to the pInvoke signature and call.
+
+### Properties and fields
+
+A property has to be marshalled as up to three different things:
+
+- A field (stored properties on frozen structs)
+- A PInvoke
+- A property
+
+A simple handler can be created which will do the necessary work and update the necessary bits of the representation.
+
+**NOTE**: Each of the handlers will need to decide whether a given method or property is supported or not. This can be done by checking the TypeDatabase for the appropriate type. This can be done by adding a separate flag to the MethodRepresentation or by having some handler validate the arguments / properties types beforehand.
```

Comment:
> Each of the handlers will need to decide whether a given method or property is supported or not.

Since we are working with pruned graph here, we should be confident that there is always a handler for each encountered type, correct?

Reply:
There are two stages of pruning necessary:
1. Pruning types - this is done before this step.
2. Pruning unknown type members - e.g there is a generic constructor on a type. This is done here.

What I mean is that at any given point of this project development up to completion there will be a set of unsupported constructs. Be it generic constructors or async properties. We need some mechanism to not emit them.

Reply:
Thanks, I didn't realize the two stages of pruning.

Do you think we could unify the stages to a single one? I.e., in the pruning graph stage also figure out the unknown type members?

Reply:
In this proposal we freeze the type database after the first pruning as it prunes types depending on whether we are able to represent their layout. The second step doesn't touch the type database - it just decided that some parts of the types should be omitted - however those pieces do not affect other constructs.

It is possible to merge those two steps into one - but then we would need to handle the types / members in topological order as we cannot rely on type database being populated